File: 42-trapping-rain-water/42-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
    var left = 0, right = height.length - 1;
    var max_left = height[left], max_right = height[right];
    var result = 0;
    while(left < right){
        if(max_left <= max_right){
            left++;
            result += calWater(height,left,max_left,max_right);
            max_left = Math.max(max_left,height[left]);
        } else {
            right--;
            result += calWater(height,right,max_left,max_right);
            max_right = Math.max(max_right,height[right]);
        }
    }

    return result;
};

function calWater(height,index,max_left,max_right){
    var water = Math.min(max_left,max_right) - height[index];
    return water > 0 ? water : 0;
}

File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        length = len(height)
        left = [-1] * length
        left[0] = height[0]
        right = [-1] * length
        right[-1] = height[-1]
        for i in range(1,length):
            left[i] = max(left[i-1],height[i])
        for i in range(length - 2,-1,-1):
            right[i] = max(right[i+1],height[i])
        result = 0
        for i in range(length):
            a = min(left[i],right[i]) - height[i]
            result += max(a,0)
        return result